File: 1917-maximum-average-pass-ratio/maximum-average-pass-ratio.py
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        # Function to compute gain if we add one student
        def gain(p, t):
            return (p + 1) / (t + 1) - p / t

        # Build max-heap (Python has min-heap, so store negative gain)
        pq = [(-gain(p, t), p, t) for p, t in classes]
        heapq.heapify(pq)

        # Distribute extra students
        while extraStudents > 0:
            g, p, t = heapq.heappop(pq)  # get class with max gain
            p, t = p + 1, t + 1
            heapq.heappush(pq, (-gain(p, t), p, t))
            extraStudents -= 1

        # Compute final average ratio
        total = 0
        for _, p, t in pq:
            total += p / t
            
        return total / len(classes)

File: 1917-maximum-average-pass-ratio/maximum-average-pass-ratio.java
class Solution {
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        // PriorityQueue: max-heap ordered by the "gain" (increase in pass ratio if we add a student)
        PriorityQueue<double[]> pq = new PriorityQueue<>(new Comparator<double[]>() {
            public int compare(double[] a, double[] b) {
                if (a[0] < b[0])
                    return 1; // bigger gain comes first
                if (a[0] > b[0])
                    return -1;
                return 0;
            }
        });

        // Initialize heap: each class contributes [gain, pass, total]
        for (int i = 0; i < classes.length; i++) {
            double pass = classes[i][0];
            double total = classes[i][1];
            // gain = new ratio - old ratio
            double gain = (pass + 1.0) / (total + 1.0) - pass / total;
            pq.offer(new double[] { gain, pass, total });
        }

        // Distribute extra students one by one
        while (extraStudents > 0) {
            double[] top = pq.poll(); // take the class with the largest gain
            double pass = top[1] + 1; // add one passing student
            double total = top[2] + 1;
            double gain = (pass + 1.0) / (total + 1.0) - pass / total;
            pq.offer(new double[] { gain, pass, total });
            extraStudents--;
        }

        // Calculate final average pass ratio
        double sum = 0.0;
        for (double[] c : pq)
            sum += c[1] / c[2];

        return sum / classes.length;
    }
}